Servers/UI/OJS.Servers.Ui/ClientApp/src/components/search/SearchSection.tsx

```diff
@@ -0,0 +1,172 @@
+import React, { useCallback, useEffect, useState } from 'react';
+import first from 'lodash/first';
+import isEmpty from 'lodash/isEmpty';
+import isNil from 'lodash/isNil';
+
+import { IContestSearchType, IProblemSearchType, IUserSearchType } from '../../common/search-types';
+import {
+    IGenericSearchProps,
+    IGenericSearchResponseModel,
+    IPagedResultType,
+} from '../../common/types';
+import { IGetSearchResultsParams } from '../../common/url-types';
+import { IErrorDataType, useHttp } from '../../hooks/use-http';
+import { useSearch } from '../../hooks/use-search';
+import { getSearchResultsUrl } from '../../utils/urls';
+import Heading, { HeadingType } from '../guidelines/headings/Heading';
+import List, { Orientation } from '../guidelines/lists/List';
+import PaginationControls from '../guidelines/pagination/PaginationControls';
+
+import styles from './SearchSection.module.scss';
+
+type ISearchTypes = IUserSearchType | IContestSearchType | IProblemSearchType;
+
+const SearchSection = <T extends ISearchTypes>({
+    searchTerm,
+    searchCategory,
+    renderItem,
+} : IGenericSearchProps<T>) => {
+    const { actions: { setSearchingError } } = useSearch();
+    const [ isLoading, setIsLoading ] = useState(false);
+    const [ searchedItems, setSearchedItems ] = useState([] as T[] || null);
+    const [ searchResults, setSearchResults ] = useState({
+        pageNumber: 1,
+        itemsPerPage: 0,
+        pagesCount: 0,
+        totalItemsCount: 0,
+    });
+    const [ itemsSearchError, setItemsSearchError ] = useState<IErrorDataType | null>(null);
+    const [ getSearchResultsParams, setGetSearchResultsParams ] =
+    useState<IGetSearchResultsParams | null>();
+    const [ currentItemsPage, setCurrentItemsPage ] = useState<number>(1);
+
+    const {
+        get: getItemSearchResults,
+        data: itemSearchResultsData,
+        error: getItemSearchResultsError,
+    } = useHttp<IGetSearchResultsParams, IPagedResultType<IGenericSearchResponseModel<T>>>({
+        url: getSearchResultsUrl,
+        parameters: getSearchResultsParams,
+    });
+
+    const setQueryParams = useCallback(
+        () => {
+            setGetSearchResultsParams({
+                searchTerm,
+                page: currentItemsPage,
+                searchCategory,
+            });
+        },
+        [ currentItemsPage, searchCategory, searchTerm ],
+    );
+
+    useEffect(() => {
+        setQueryParams();
+    }, [ setQueryParams, searchTerm, currentItemsPage ]);
+
+    useEffect(() => {
+        if (isNil(getSearchResultsParams)) {
+            return;
+        }
+
+        setIsLoading(true);
+        setItemsSearchError(null);
+        getItemSearchResults();
```

Comment:
reorder setters first then function invocation or reverse for better readability

Reply:
Setters are rendered to follow certain logic for tracking the loading of the data